File: src/com/easy/GuessNumberHigherOrLower_374.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class GuessNumberHigherOrLower_374 {
    public int guess(int num) {
        return 0;
    }

    // 2019年7月17日
    // Use binary search
    // Time Complexity: O(logn)
    // Space Complexity:O(1)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 32.9 MB, less than 100.00%
    public int guessNumber(int n) {
        int mLeft = 1, mRight = n;
        int mResult = 0;

        while (mLeft <= mRight) {
            int mMid = mLeft + (mRight - mLeft) / 2;
            if (guess(mMid) == -1) {
                mRight = mMid - 1;
            } else if (guess(mMid) == 1) {
                mLeft = mMid + 1;
            } else {
                mResult = mMid;
                break;
            }

        }

        return mResult;
    }
}

File: Golang_src/easy/guessNumberHigherOrLower_374.go
package easy

// Use binary search:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 1.9 MB, less than 88.15%
func guessNumber(n int) int {

	// mock function
	guess := func(num int) int {
		return num
	}

	l, r := 0, n

	for l <= r {
		mid := l + (r-l)/2
		switch guess(mid) {
		case -1:
			r = mid - 1
		case 1:
			l = mid + 1
		case 0:
			return mid
		}
	}

	return -1
}
